### PR TITLE
Make the config's `paths` field more extensible and implement its resolution logic

### PR DESCRIPTION
**Pull request by alcuadrado**
This PR completes several TODOs around the config's `paths` field.

This is a summary of the changes of this PR:

1. Changes its types a bit to make it accept an array of strings where it makes sense and to make the `tests` and `sources` fields extensible through plugins.

2. It updates the `mocha` and `node:test` plugins to take advantage of that, defining different config fields to avoid clashing with each other.

3. It adds the logic to resolve the default `paths` of `hardhat`, and those of the `mocha` and `node:test` plugins.

4. It introduces a `solidity` builtin plugin, which just resolves a config field for now.

5. It implements validation of the config `paths` of the plugins affected by this PR, but not the rest of the `paths` in general.

In a future PR we should address these things:

- [hardhat] Validate the rest of the config `paths`
- [hardhat] Test the config `paths` resolution
- [hardhat-mocha] Test the config `paths` resolution
- [hardhat-node-test-runner] Test the config paths resolution
- [builtin:solidity] Test the config paths resolution

### CODE REVIEW

**Review by schaable — COMMENTED** [review]
Looks good! I left a couple comments regarding hardhat-utils helpers.

File: v-next/hardhat/src/internal/builtin-plugins/solidity/index.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import type { HardhatPlugin } from "../../../types/plugins.js";
+
+import "./type-extensions.js";
+
+const hardhatPlugin: HardhatPlugin = {
+  id: "builtin:solidity",
```
[the remark this answers is not in the document]
Comment (alcuadrado) [reply v-next/hardhat/src/internal/builtin-plugins/solidity/index.ts:6]: yes, we can upadte all of them on a separate pr

File: v-next/hardhat-mocha-test-runner/src/hookHandlers/config.ts [hunk 2]
```diff
@@ -88,13 +94,29 @@ export default async (): Promise<Partial<ConfigHooks>> => {
         resolveConfigurationVariable,
       );
 
+      let testsPath = userConfig.paths?.tests;
+
+      testsPath = typeof testsPath === "object" ? testsPath.mocha : testsPath;
```
[the remark this answers is not in the document]
Comment (schaable) [reply v-next/hardhat-mocha-test-runner/src/hookHandlers/config.ts:99]: We need to find a way to make `isObject` narrow the type correctly. In the meantime, I've added TODOs to remind us to use `isObject` once the type narrowing behaves as expected.

**Review by schaable — APPROVED** [review]
@alcuadrado I've updated the code with the changes we discussed. We still need to fix the `isObject` helper, but we can address that in a separate PR.

**Pull request merged**